### ISSUE TITLE
Invalid memory access (signal 11) when invoking a stored Proc from a C callback

### ISSUE THREAD

**Issue by furkanmustafa** [crystal-lang/crystal#2666]
After issue #1973 has been fixed, I have updated my code and kept working on it a bit.
This time, I have managed to utilize a C callback, avoiding Closures, using Box etc. etc. But invoking a Proc that I have stored in an instance variable crashes. Even though it looks like I'm successfully unboxing my instance. Might be my mistake, but couldn't figure out yet.

**Full Code:** (2 files) https://gist.github.com/furkanmustafa/4e07a7216716c0a78d6d

**Partial Code:** (the part that it invokes the error below)
```cr
  class Timer
    def initialize(_loop : UV::Loop)
      _loop = Loop.default unless _loop
      @timer = LibUV::UvTimerS.new
      LibUV.timer_init(_loop._underlyingloop, pointerof(@timer))
      @timer.data = Box.box(self)
      puts "Boxed our timer: #{@timer.data.to_s} into UvTimer(#{pointerof(@timer).to_s})"
    end

    def run_action
      puts "Calling the! action"
      if action = @action
        action.call # <<< POINT OF CRASH
      end
    end

    def start(timeout, repeat, &action)
      puts "Starting timer"
      @action = action
      proc = ->(_handle : LibUV::UvTimerS*, status : LibC::Int) {
        puts "Reviving underlying timer from #{_handle.to_s}"
        _uv_timer = _handle.value
        puts "Reviving our timer: #{_uv_timer.to_s}.data = #{_uv_timer.data.to_s}"
        timer_instance = Box(UV::Timer).unbox(_uv_timer.data)
        puts "Calling run_action"
        timer_instance.run_action
      }
      LibUV.timer_start(pointerof(@timer), proc, timeout, repeat)
    end

    def stop
      LibUV.timer_start(pointerof(@timer))
    end

    def again
      LibUV.timer_again(pointerof(@timer))
    end
  end
```

**Output:**

```
$ crystal main.cr 
Will quit in a bit
Boxed our timer: Pointer(Void)@0x1113f20 into UvTimer(Pointer(LibUV::UvTimerS)@0x111df08)
Starting timer
(waits 5 seconds here)
Reviving underlying timer from Pointer(LibUV::UvTimerS)@0x111df08
Reviving our timer: LibUV::UvTimerS(@data=Pointer(Void)@0x1113f20, @loop=Pointer(LibUV::UvLoopT)@0x7f2baa1bc900, @type=UvTimer, @close_cb=#<(Pointer(LibUV::UvHandleT) -> Void):0x0>, @handle_queue=LibUV::NgxQueueT(), @flags=0, @next_closing=Pointer(LibUV::UvHandleT).null, @tree_entry=LibUV::UvTimerSTreeEntry(@rbe_left=Pointer(LibUV::UvTimerS).null, @rbe_right=Pointer(LibUV::UvTimerS).null, @rbe_parent=Pointer(LibUV::UvTimerS).null, @rbe_color=24576), @timer_cb=#<(Pointer(LibUV::UvTimerS), Int32 -> Void):0x41d500>, @timeout=0, @repeat=0, @start_id=0).data = Pointer(Void)@0x1113f20
Calling run_action
Calling the! action
Invalid memory access (signal 11) at address 0x13e910b2
[4321573] *CallStack::print_backtrace:Int32 +117
[4306376] __crystal_sigfault_handler +56
Program exited because of a segmentation fault (11)
```

Crystal 0.17.3 (Fri May 20 19:28:36 UTC 2016), on Arch Linux, Linux 4.6.0

**Comment by furkanmustafa** [comment]
Just confirmed that problem occurs at `0.17.4 [032ab91]` too.

**Comment by jhass** [comment]
That does in fact look like a codegen bug, the following diff works around:

```diff
diff --git a/libuv.cr b/libuv.cr
index e269438..3567a07 100644
--- a/libuv.cr
+++ b/libuv.cr
@@ -664,7 +664,9 @@ module UV
   end
 
   class Timer
+    @action : ->
     def initialize(_loop : UV::Loop)
+      @action = -> { Pointer(Void).null.value }
       _loop = Loop.default unless _loop
       @timer = LibUV::UvTimerS.new
       LibUV.timer_init(_loop._underlyingloop, pointerof(@timer))
```

Perhaps related, doing just `@action = -> { }` throws
```
in ./libuv.cr:669: instance variable '@action' of UV::Timer must be ( -> Void), not ( -> Nil)

      @action = -> { }
```

and that with `@action : -> Void?` throws

```
in ./libuv.cr:669: instance variable '@action' of UV::Timer must be ( -> Void?), not ( -> Nil)

      @action = -> { }
```


On another note, you are aware that Crystal already includes libevent and uses it for all stdlib IO operations? For example to port your example

```cr
spawn do
  loop do
    sleep 5
    puts "Wow timer"
  end
end

sleep
```

**Comment by ysbaddaden** [comment]
The local `proc` variable is passed to C but never retained in Crystal code: the GC will collect it. No?

**Comment by jhass** [comment]
My first thought too, but no. the proc literal generates a function and merely a pointer to that function is passed, the local var is eliminated.

**Comment by asterite** [comment]
Whenever you get a segfault, one first has to check that the bindings are correct. A wrong binding for a struct, for example missing to map some fields, will result in some parts of the stack being corrupted.

Compiling a C program and checking `sizeof(struct uv_timer_s)` gives 152, but then sizeof(LibUV::UvTimerS) gives 128. I added this:

```crystal
    more : Int8[24]
```

to the struct definition and it started working well.

Why are you binding to libuv? Crystal already has non-blocking IO. It uses `libevent`, but that could change in the future.

**Comment by asterite** [comment]
Oh, I forgot to mention that if you print `@action` inside `Timer#run_action` for me it said:

```
#<( -> Void):0x4b12676:closure>
```

but printing `@action` after it was set I got:

```
#<( -> Void):0x106d67480>
```

so it seems `@action` changed. That almost always means the stack got corrupted, somehow. It could mean a compiler bug too, but that wasn't the case here :-)

**Comment by furkanmustafa** [comment]
Ouch. @asterite thanks! Then it's probably my libuv version has changed drastically since when I have prepared that binding file. Or I've messed up badly when generating it.

I can confirm that segfault doesn't occur anymore. There is an issue of callback getting called infinitely when the timer fires, but that's another issue that I need to study and understand first.

Btw, for the reason I'm messing with libuv ; I'm just trying to play with crystal, bindings, and also trying to get libuv's event loop running here.

For the issue; `libuv.cr:669: instance variable '@action' of UV::Timer must be ( -> Void?), not ( -> Nil)` that doesn't happen for me at the moment, but happened a few times on several other places. It's probably a different issue.

Closing the issue, thanks, and sorry for my mistake there.

**Comment by jhass** [comment]
@asterite did I miss something in the new type inference rules that still allows this to compile?

```cr
class Foo
  def bar(&block)
    @action = block
  end
  
  def baz
    if action = @action
      action.call
    end
  end
end
  
foo = Foo.new
foo.bar do
  puts "hi"
end
foo.baz
```
https://carc.in/#/r/103l

Also what about the other (two) issue(s) I mentioned?
https://carc.in/#/r/103m and https://carc.in/#/r/103n, shouldn't Void vanish/accept everything?

**Comment by asterite** [comment]
@jhass The last snippet is fine, I added that rule in the last version. The compiler knows `block` doesn't have arguments and returns void, so it infers that.

Then, https://carc.in/#/r/103m should compile, and I actually faced this a couple of times. I guess merging proc types where one returns void and the other doesn't, with the same argument types, should give a proc that returns void.

For https://carc.in/#/r/103n, not sure. `Void?` looks strange.

**Comment by asterite** [comment]
Oh, the `Void?` shouldn't compile, the correct type would be `(-> Void) | (-> Nil)`.

Another idea I had these days is to remove `Void` from the language. `Nil` would take that meaning. We could use `alias Void = Nil` for the upgrade process.

Basically, `Void` is a type that behaves different than other types: you can't inspect it, you can't use it, it's not very intuitive. Methods retuning `nil`, in the codegen, would generate functions that return `void`. lib calls returning `Void` would now return `Nil`, meaning "nothing is returned", but you can print it. For example, right now you can't do:

```crystal
lib LibFoo
  fun foo
end

def print_something(obj)
  puts obj
end

print_something 1 # ok
print_something LibFoo.foo # error, undefined method 'to_s' for Void
```

you have to use an `if is_a?(Void)` as a special case.

Basically, `Void` belongs to C. In a language where there are no statements and everything is an expression and everything has a value, maybe `Void` shouldn't be a type.

But I'm still not 100% convinced about this.

**Comment by jhass** [comment]
I actually like that idea very much, `Void` always felt "not right" to me. If the method it returns something, it's clear that `nil` is a null pointer, if it's a void method, well, what do you care what `nil` means here.

**Comment by asterite** [comment]
@jhass Exactly! `nil` basically means `I don't care about this value`.

I'm thinking, what bothers me about having `Proc(Void)` and then having `->{ ... }` not be compatible is that there's no way to easily produce a void value. I can't do `->{ void }`. However, if `Void` disappears and is replaced with `Nil`, I can simply do `->{ ... ; nil }`, and it's pretty explicit that I don't care about the return value. However, I'd still like to consider #2672 in this, because it's obvious that the instance variable has a proc that doesn't care about the return type, so assigning a proc that does care about the return type should be implicitly convertible.

**Comment by asterite** [comment]
I'll open a separate RFC for this. I think it can simplify the language and the compiler too.

**Comment by asterite** [comment]
@furkanmustafa No worries, binding to big libraries like this is always a headache. We'd like to improve this in Crystal, we still don't know how.

**Comment by tjgillies** [comment]
> for example missing to map some fields, will result in some parts of the stack being corrupted.

This just helped me a lot heh. So much for being lazy
